src/lib.rs

```diff
@@ -251,110 +247,116 @@ impl<'a, T> core::clone::Clone for Menu<'a, T> {
-                }
-            }
-            // Handle the command
-            self.process_command();
-            Outcome::CommandProcessed
-        } else if (input == 0x08) || (input == 0x7F) {
-            // Handling backspace or delete
-            if self.used > 0 {
-                write!(self.context, "\u{0008} \u{0008}").unwrap();
-                self.used -= 1;
+    pub fn process(&mut self, context: &mut T) {
+        while context.read_ready().unwrap() {
+            let mut input_buf = [0; 1];
+            context.read(&mut input_buf).unwrap();
+            let input = input_buf[0];
+
+            // Strip carriage returns
+            if input == 0x0A {
+                continue;
             }
-            Outcome::NeedMore
-        } else if self.used < self.buffer.len() {
-            self.buffer[self.used] = input;
-            self.used += 1;
 
-            #[cfg(feature = "echo")]
-            {
-                // We have to do this song and dance because `self.prompt()` needs
-                // a mutable reference to self, and we can't have that while
-                // holding a reference to the buffer at the same time.
-                // This line grabs the buffer, checks it's OK, then releases it again
-                let valid = core::str::from_utf8(&self.buffer[0..self.used]).is_ok();
-                // Now we've released the buffer, we can draw the prompt
-                if valid {
-                    write!(self.context, "\r").unwrap();
-                    self.prompt(false);
+            let outcome = if input == 0x0D {
+                #[cfg(not(feature = "echo"))]
+                {
+                    // Echo the command
+                    write!(context, "\r").unwrap();
+                    if let Ok(s) = core::str::from_utf8(&self.buffer[0..self.used]) {
+                        write!(context, "{}", s).unwrap();
+                    }
                 }
-                // Grab the buffer again to render it to the screen
-                if let Ok(s) = core::str::from_utf8(&self.buffer[0..self.used]) {
-                    write!(self.context, "{}", s).unwrap();
+                // Handle the command
+                self.process_command(context);
+                Outcome::CommandProcessed
+            } else if (input == 0x08) || (input == 0x7F) {
+                // Handling backspace or delete
+                if self.used > 0 {
+                    write!(context, "\u{0008} \u{0008}").unwrap();
+                    self.used -= 1;
                 }
+                Outcome::NeedMore
+            } else if self.used < self.buffer.len() {
+                self.buffer[self.used] = input;
+                self.used += 1;
+
+                #[cfg(feature = "echo")]
+                {
+                    // We have to do this song and dance because `self.prompt()` needs
+                    // a mutable reference to self, and we can't have that while
```

Comment:
Is this still true now context is passed out of band?

Reply:
Yes this is still true because of borrowing `self.buffer` for the string and then mutably borrowing `self` when calling `self.prompt()`

Reply:
ahh

Reply:
but can prompt be made non-mut?